File: go/problem_36.go
package main

import "fmt"

// BSTNode is a binary search tree node.
type BSTNode struct {
	val   int
	left  *BSTNode
	right *BSTNode
}

// secondLargest finds the second-largest value in a BST.
// The largest is the rightmost node. The second largest is either:
//   (a) the largest node in the left subtree of the rightmost node, or
//   (b) the parent of the rightmost node (if it has no left child).
func secondLargest(root *BSTNode) int {
	var parent, curr *BSTNode
	curr = root
	for curr.right != nil {
		parent = curr
		curr = curr.right
	}
	// curr is the largest node
	if curr.left != nil {
		// second largest = rightmost of curr.left
		node := curr.left
		for node.right != nil { node = node.right }
		return node.val
	}
	return parent.val
}

func problem36() {
	fmt.Println("Problem 36:")
	//       5
	//      / \
	//     3   7
	//    / \ / \
	//   2  4 6  8
	root := &BSTNode{5,
		&BSTNode{3, &BSTNode{2, nil, nil}, &BSTNode{4, nil, nil}},
		&BSTNode{7, &BSTNode{6, nil, nil}, &BSTNode{8, nil, nil}},
	}
	if secondLargest(root) != 7 { panic("assertion failed") }
	// Tree: 3 -> 4 -> 5
	root2 := &BSTNode{3, nil, &BSTNode{4, nil, &BSTNode{5, nil, nil}}}
	if secondLargest(root2) != 4 { panic("assertion failed") }
	// Tree: 5 with left subtree max = 4
	root3 := &BSTNode{5, &BSTNode{3, nil, &BSTNode{4, nil, nil}}, nil}
	if secondLargest(root3) != 4 { panic("assertion failed") }
	fmt.Println("All Tests Passed")
}

File: python/problem_36.py
class BSTNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def second_largest(root: BSTNode) -> int:
    parent, curr = None, root
    while curr.right:
        parent = curr
        curr = curr.right
    if curr.left:
        node = curr.left
        while node.right:
            node = node.right
        return node.val
    return parent.val


def main():
    print("Problem 36:")
    root = BSTNode(5,
        BSTNode(3, BSTNode(2), BSTNode(4)),
        BSTNode(7, BSTNode(6), BSTNode(8))
    )
    assert second_largest(root) == 7
    root2 = BSTNode(3, None, BSTNode(4, None, BSTNode(5)))
    assert second_largest(root2) == 4
    root3 = BSTNode(5, BSTNode(3, None, BSTNode(4)), None)
    assert second_largest(root3) == 4
    print("All Tests Passed")
